### PR TITLE
Include Implicit Trapezoidal for time stepping

### PR DESCRIPTION
**Pull request by bindi-nagda**

### CODE REVIEW

File: INSVCTwoFluidStaggeredHierarchyIntegrator.cpp [hunk 1]
```diff
@@ -494,10 +497,36 @@ INSVCTwoFluidStaggeredHierarchyIntegrator::integrateHierarchy(const double curre
...
+            f2_vec->copyVector(Pointer<SAMRAIVectorReal<NDIM, double>>(&f_vec, false), true);
+            stokes_op.initializeOperatorState(u_vec, *f2_vec);
+            stokes_op.apply(u_vec, *f2_vec);
+
+            // Sum f_vec and f2_vec and store result in f_vec
+            Pointer<SAMRAIVectorReal<NDIM, double>> f_vec_ptr(&f_vec, false);
+            f_vec.axpy(1.0, f_vec_ptr, f2_vec, true);
+
+        case BACKWARD_EULER:
+            const double D2 = -1.0; // This will depend on the time stepping scheme.
```
Comment (abarret) [inline INSVCTwoFluidStaggeredHierarchyIntegrator.cpp:524]: Suggest creating two operators, a "solve" operator and a "rhs" operator. Only set the `D1` and `D2` (and the `C` coefficients if necessary) in the switch statement. Then apply the operator outside the switch statement. This way you can remove the for loop above. You also won't have branching code for each time stepping type.

File: INSVCTwoFluidStaggeredHierarchyIntegrator.cpp [hunk 2]
```diff
@@ -494,10 +497,36 @@ INSVCTwoFluidStaggeredHierarchyIntegrator::integrateHierarchy(const double curre
...
+    }
+    
     // Setup the stokes operator
     Pointer<VCTwoFluidStaggeredStokesOperator> stokes_op = new VCTwoFluidStaggeredStokesOperator("stokes_op", true);
-    const double D = -1.0; // This will depend on the time stepping scheme.
-    stokes_op->setCandDCoefficients(C, D);
+    stokes_op->setCandDCoefficients(C, D2);
```
Comment (abarret) [inline INSVCTwoFluidStaggeredHierarchyIntegrator.cpp:529]: Variable `D2` is out  of scope and can't be used here.

File: INSVCTwoFluidStaggeredHierarchyIntegrator.cpp (hunk 1 above)
Comment (bindi-nagda) [reply INSVCTwoFluidStaggeredHierarchyIntegrator.cpp]: I pushed some new changes based on your review. You mentioned creating a "solve" operator, but I wasn't sure what you wanted for that.

File: INSVCTwoFluidStaggeredHierarchyIntegrator.cpp [hunk 3]
```diff
@@ -456,48 +459,48 @@ INSVCTwoFluidStaggeredHierarchyIntegrator::integrateHierarchy(const double curre
...
-                double thn_lower = 0.5 * ((*thn_data)(idx_c_low) + (*thn_data)(idx_c_up)); // thn(i,j-1/2)
-                (*f_un_data)(idx) = (*f_un_data)(idx) + C * thn_lower * (*un_data)(idx);
-                (*f_us_data)(idx) = (*f_us_data)(idx) + C * convertToThs(thn_lower) * (*us_data)(idx);
-            }
-        } // patches
-    }     // levels
+    // set-up RHS to treat viscosity and drag with backward Euler or Implicit Trapezoidal Rule: 
+    // RHS = f(n) + C*theta_i(n)*u_i(n) + D1*(pressure + viscous + drag) for  i = n, s
+    const double D1;
+    const double D2;
```
Comment (abarret) [inline INSVCTwoFluidStaggeredHierarchyIntegrator.cpp:465]: These should not be `const`.
```suggestion
    double D1;
    double D2;
```

File: INSVCTwoFluidStaggeredHierarchyIntegrator.cpp [hunk 4]
```diff
@@ -456,48 +459,48 @@ INSVCTwoFluidStaggeredHierarchyIntegrator::integrateHierarchy(const double curre
...
+            break;
+
+        case BACKWARD_EULER:
+            D1 = 0.0;
+            D2 = -1.0; 
+            break;
+
+        default:  // Use Backward Euler
+            D1 = 0.0;
+            D2 = -1.0;
```
Comment (abarret) [inline INSVCTwoFluidStaggeredHierarchyIntegrator.cpp:480]: Let's quit with a helpful error describing the problem instead.
```suggestion
            TBOX_ERROR("Unknown time stepping type " + string_to_num(d_viscous_time_stepping_type) + ". Valid options are BACKWARD_EULER and TRAPEZOIDAL_RULE.");
```

**Review by abarret — CHANGES_REQUESTED** [review]
Two more small changes. It might be worth adding the ability set different `C` coefficients so  that we could use BDF schemes, but we can do that later.

Did you compile and run this to make sure it works?

**Comment by bindi-nagda** [comment]
> Did you compile and run this to make sure it works?

Yes, compilation and execution was successful for both time stepping types.

**Review by abarret — APPROVED** [review]
:+1:

**Pull request merged**